### ISSUE TITLE
Set map bounds to circle bounds? (polyline, polygon, marker, circle)

### ISSUE THREAD

**Issue by ollietreend** [angular-ui/angular-google-maps#449]
Hi,

**Intro**
I am currently building a search tool which will plot search results on a map, similar to your typical store locator. The user chooses a location (lat/lon) to search from and a radius. A database is queried for records which are within the specified radius of the user's location. These results are returned back to the browser to be plotted on a Google Map with angular-google-maps.

I am using `<markers>` directive to display the results on the map. I am also using a `<circle>` directive to show the search radius, setting the `"center"` property to be the user's chosen location, and the `"radius"` property to be the search radius.

**The Problem**
I would like to have the map fit to the bounds of the circle. I understand that I can set the `"bounds"` property on the `<google-map>` directive. However I'm not sure how I can retrieve the bounds of the circle.

**Potential Solution**
The [google.maps.Circle class](https://developers.google.com/maps/documentation/javascript/reference#Circle) has the method `getBounds()` which returns a [LatLngBounds object](https://developers.google.com/maps/documentation/javascript/reference#LatLngBounds). However I'm not sure if I'm able to access the `Circle` object of the `<circle>` directive in order to retrieve its bounds.

Even if I am able to retrieve the bounds of the `<circle>`, I'm not sure how I would store that in a scope model in order to assign it to the `<google-map>`.

Do you know how I could achieve the above using angular-google-maps?

**Also Tried**
I have also tried setting `"fit"` to `true` on the `<markers>` directive. This works as expected, however because it fits the map based on the returned results, it doesn't center the map based on the initial search location, nor does it fit to the radius circle.

I want to show the search location in the center, even if the results only exist within the southern region of the radius circle. Therefore this approach doesn't achieve what I'm looking for.

Thanks

**Comment by Jerome2606** [comment]
So, is there any enhancements for this feature?

**Comment by nmccready** [comment]
No but I just added it to the milestone list. Feel free to have at it.

**Comment by ollietreend** [comment]
OP here. The original project is long gone now (as of about 1 year), and I'm trying to remember how I worked around this (in case it's of any help to @Jerome2606). IIRC I did something along the lines of – centre the map based on the initial search lat/lon, and then 'guesstimated' the appropriate zoom level based on the user's chosen search radius.

The map in question needed to be responsive down to mobile devices, so unfortunately that added another variable into the mix. I think I played around with changing the zoom level based on screen width (`window.matchMedia` came in handy here), but that made things rather complicated so I gave up on that and stuck with 1 zoom level per radius.

YMMV, but that's the solution I went with. (Again, IIRC.)

Of course a proper solution to this would be ideal, but this doesn't seem to be a highly requested feature.

**Comment by Jerome2606** [comment]
I found a way to get Circle:

Use control attribute to circle directive and it's populate an object with function getCircle:
```
<ui-gmap-circle events="map.circle.events" control="map.circle.gObject"></ui-gmap-circle>
```

After you can simply do:
```
$scope.map.circle = {
    ...
    gObject: {},
                    events:{
                        radius_changed: function(){
                            
                            $scope.map.control.getGMap().fitBounds($scope.map.circle.gObject.getCircle().getBounds());
                        }
                    }
};
```

**Comment by ollietreend** [comment]
@Jerome2606 perfect! It looks like the control object on the circle directive is new – that wasn't possible when I opened this ticket.

A 'fit' attribute on the circle directive would be the ideal solution, but the use of the control object is a nice workaround.

**Comment by nmccready** [comment]
Yeah I'll add fit since im in the midst of adding it to polyline, polygon, and marker.

**Comment by dmackerman** [comment]
I'm running into a similar issue, but what I'm trying to do is use `fitBounds` method on my map (or the `bounds` config) to fit the map bounds to an array of `Circles` that are being drawn with the `ui-gmap-circle` directive.

Is there anyway to access the raw, underlying `google.maps.Circle` object for each of these after they are drawn on the map? I can figure out the bounding logic quite easily if I had access to the `getBounds()` methods on the `Circle`'s themselves.

Pretty standard layout here:

```html
                  <ui-gmap-google-map
                      center="detail.map.center"
                      options="detail.mapOptions"
                      zoom="detail.map.zoom"
                      events="detail.mapEvents">
                      <ui-gmap-circle
                          id="zone.id"
                          ng-repeat="zone in detail.profile.zones"
                          center='zone.center'
                          radius='zone.radius'
                          fill='zone.fill'
                          stroke='zone.stroke'
                          name="zone.name"
                          clickable='true'
                          draggable='true'
                          editable='true'
                          visible='true'
                          events='detail.zoneEvents'
                          >
                      </ui-gmap-circle>
                  </ui-gmap-google-map>
```

I've tried inspecting the `map` instance itself and it doesn't appear that anything is being added in terms of seeing which shapes are present on the map.

Any ideas on how I can access the raw `Circle` instances?

**Comment by blaues0cke** [comment]
After a lot of sturggle, I came up with a modified version of the solution from @Jerome2606, maybe it will help someone to workaround this:

    <ui-gmap-circle
        ...
        events="mapObjects.circle.events"
    >
    </ui-gmap-circle>

+

    events:{
        dragend: function(circle, eventName, scope) {
            circle.getMap().fitBounds(circle.getBounds());
        },
        radius_changed: function(circle, eventName, scope) {
            circle.getMap().fitBounds(circle.getBounds());
        }
    }

This will automatically zoom to the cricles bounds.

For polygons, this should work for dragging:

    <ui-gmap-polygon
        ...
        events="mapObjects.polygon.events"
    >
    </ui-gmap-polygon>

+

    events:{
        dragend: function(polygon, eventName, scope) {
            var bounds = new google.maps.LatLngBounds();
            var paths = polygon.getPaths();
            var path;
            for (var i = 0; i < paths.getLength(); i++) {
                path = paths.getAt(i);
                for (var ii = 0; ii < path.getLength(); ii++) {
                    bounds.extend(path.getAt(ii));
                }
            }

            polygon.getMap().fitBounds(bounds);
        }
    }

**Comment by ROJAHEMA** [comment]
Hi,

Is there any feature for drawing curved /Arc polyline between two latitude and longitude.
If so can you please give me some info...?
